lib/puppet/network/resolver.rb

```diff
@@ -1,85 +1,134 @@
 require 'resolv'
-module Puppet::Network; end
-
-module Puppet::Network::Resolver
-  # Iterate through the list of servers that service this hostname
-  # and yield each server/port since SRV records have ports in them
-  # It will override whatever masterport setting is already set.
-  def self.each_srv_record(domain, service_name = :puppet, &block)
-    if (domain.nil? or domain.empty?)
-      Puppet.debug "Domain not known; skipping SRV lookup"
-      return
-    end
 
-    Puppet.debug "Searching for SRV records for domain: #{domain}"
+module Puppet::Network
+  class Resolver
+
+    def initialize
+      @resolver = Resolv::DNS.new
 
-    case service_name
-      when :puppet then service = '_x-puppet'
-      when :ca     then service = '_x-puppet-ca'
-      when :report then service = '_x-puppet-report'
-      when :file   then service = '_x-puppet-fileserver'
-      else              service = "_x-puppet-#{service_name.to_s}"
+      # Stores DNS records per service, along with their TTL
+      # and the time at which they were resolved, for cache
+      # eviction.
+      @record_cache = {}
     end
-    srv_record = "#{service}._tcp.#{domain}"
-
-    resolver = Resolv::DNS.new
-    records = resolver.getresources(srv_record, Resolv::DNS::Resource::IN::SRV)
-    Puppet.debug "Found #{records.size} SRV records for: #{srv_record}"
-
-    if records.size == 0 && service_name != :puppet
-      # Try the generic :puppet service if no SRV records were found
-      # for the specific service.
-      each_srv_record(domain, :puppet, &block)
-    else
-      each_priority(records) do |priority, recs|
-        while next_rr = recs.delete(find_weighted_server(recs))
-          Puppet.debug "Yielding next server of #{next_rr.target.to_s}:#{next_rr.port}"
-          yield next_rr.target.to_s, next_rr.port
+
+    # Iterate through the list of servers that service this hostname
+    # and yield each server/port since SRV records have ports in them
+    # It will override whatever masterport setting is already set.
+    def each_srv_record(domain, service_name = :puppet, &block)
+      if (domain.nil? or domain.empty?)
+        Puppet.debug "Domain not known; skipping SRV lookup"
+        return
+      end
+
+      Puppet.debug "Searching for SRV records for domain: #{domain}"
+
+      case service_name
+        when :puppet then service = '_x-puppet'
+        when :ca     then service = '_x-puppet-ca'
+        when :report then service = '_x-puppet-report'
```

Comment:
While we're doing this, can we get rid of the `:ca` and `:report` clauses? I think they're handled by the else clause (I'd love to get rid of the `:file` clause to, but I realize that would involve editing the callers as well and might not be worth it).

Reply:
Originally `:ca`, and `:report` were in there as more of a "note to self" about which services we were explicitly supporting, but you're right that they're covered by the `else` clause (as is anything where our internal service name matches the suffix we'd use with `_x-puppet-` to generate the SRV service name).

I'm ambivalent about whether the `:ca` and `:report` clauses get removed. I just wanted to provide context around why they're explicitly called out there.

Reply:
Yeah I could also go either way. Seeing `:ca` called out separately definitely helped me understand how this was being used when I was initial reading through the certificate fetching code.